### ISSUE TITLE
StorageSubscription.Unsubscribe race

### ISSUE THREAD

**Issue by ggwpez** [centrifuge/go-substrate-rpc-client#188]
Hi there,

i am using `gsrpc` in a project and noticed that a race occurs when trying to unsubscribe from a `StorageSubscription` while a second one is running.  
This is weird since i never spawned a go-routine myself.  

Place the following test in `main_test.go` and execute `go test -v -count=1 -race -run=TestClient_Subscribe .`:
```go
func TestClient_Subscribe(t *testing.T) {
	// Setup and chain connection
	api, err := gsrpc.NewSubstrateAPI(config.Default().RPCURL)
	require.NoError(t, err)
	meta, err := api.RPC.State.GetMetadataLatest()
	require.NoError(t, err)
	// Use "Time" if you run an older node version.
	key, err := types.CreateStorageKey(meta, "Timestamp", "Now")
	require.NoError(t, err)

	// Subscribe once
	sub1, err := api.RPC.State.SubscribeStorageRaw([]types.StorageKey{key})
	require.NoError(t, err)

	// Subscribe twice
	_, err = api.RPC.State.SubscribeStorageRaw([]types.StorageKey{key})
	require.NoError(t, err)

	// Unsubscribing produces a race!
	sub1.Unsubscribe()
}
```
Test output: [race.txt](https://github.com/centrifuge/go-substrate-rpc-client/files/7176997/race.txt).  
The problem seems to occur between https://github.com/centrifuge/go-substrate-rpc-client/blob/fee0d0ef2a142a5908e13e75c5ce51eef859dc68/gethrpc/subscription.go#L304 and https://github.com/centrifuge/go-substrate-rpc-client/blob/03ce7c572bf00722449fafbccecc1dd4fbfc4fe5/rpc/state/subscribe_storage.go#L59
Any help with this would be much appreciated, thanks for your time :pray:

**Comment by NunoAlexandre** [comment]
@ggwpez did you happen to resolve this issue?

**Comment by ggwpez** [comment]
> @ggwpez did you happen to resolve this issue?

No, we just refrained from running the race tests in [our project](https://github.com/perun-network/perun-polkadot-backend)…  
I saw that this is old code copied from [geth](https://github.com/ethereum/go-ethereum), maybe it could help to update it? I dont know.  
Instead it could also use a fork as subrepo here, to keep track which files were changed from the geth code.  
@NunoAlexandre
